src/gateio_new_coins_announcements_bot/rotating_proxy.py

```diff
@@ -0,0 +1,105 @@
+from typing import Callable
+import requests
+import threading
+import time
+import itertools
+import socket
+import struct
+
+import gateio_new_coins_announcements_bot.globals as globals
+from gateio_new_coins_announcements_bot.logger import logger
+
+
+_proxy_list = {}
+_proxy = None
+_event = threading.Event()
+
+
+def init_proxy():
+    threading.Thread(target=lambda: _every(60 * 10, _fetch_proxies)).start()
+    # Required for populating the proxy list when starting bot
+    _fetch_proxies()
+
+
+def _fetch_proxies():
+    logger.info("Fetching proxies...")
+    global _proxy_list
+    global _proxy
+    threads: list[threading.Thread] = []
+    try:
+        proxy_res = requests.get(
+            "https://www.proxyscan.io/api/proxy?last_check=180&limit=20&type=socks5&format=txt&ping=1000"
+        ).text
+    except requests.exceptions.RequestException as e:
+        logger.error(e)
+
+    # Merging old proxies with new ones
+    _list = list(proxy_res[:-1].split("\n") | _proxy_list.keys())
+
+    if len(_list) > 0:
+        for p in _list:
+            t = threading.Thread(target=checker, args=[p])
+            t.start()
+            threads.append(t)
+
+        for t in threads:
+            t.join()
+
+    logger.info(f"Fetched {len(_proxy_list)} proxies")
+    _proxy = itertools.cycle(_proxy_list.keys())
+
+
+def get_proxy() -> str:
+    return next(_proxy)
```

Comment:
Will cause a crash if _proxy is empty `next([])`, which can happen if `_fetch_proxies()` fails.
Raise a new exception from the old one (see [here](https://www.python.org/dev/peps/pep-3134/#explicit-exception-chaining)) with a meaningful error message like : "No proxies available", since it shouldn't be called unless `is_ready()` is true anyway